Implement worksheets API

kashizui opened the pull request:
Resolves #579 
Resolves #513 
Resolves #516 
Resolves #313

Sorry about the huge PR, I kind of went hog wild.

Cleaned up XMLRPC and BundleClients stuff as well.

Should be merged in after codalab/codalab-worksheets#219 because I ripped out the old bundle upload API.
- [x] ls
- [x] wls
- [x] work
- [x] finish search
- [x] finish mimic
- [x] new
- [x] wrm
- [x] wedit
- [x] detach
- [x] wperm
- [x] print
- [x] add
- [x] wadd
- [x] rebase
- [x] move remaining required LocalBundleClient functionality into legacy.BundleService.
- [x] check and debug test-cli locally
- [x] check and debug test-cli on circleci
- [x] test and debug the website
- [x] wizard to migrate users' default aliases
- [x] cleanup uuid resolution code (merge into canonicalize?)
- [x] fix completers

@percyliang (@yashsavani @ChengshuLi)

kashizui committed:
WIP

kashizui committed:
Remove CodaLabManager use_rest hacks

kashizui committed:
Clean up some use_rest stuff in CLI

kashizui committed:
More cleaning hacks and untying worksheet_util

kashizui committed:
Clean editable metadata thingy

kashizui committed:
Checkpoint for using resolver

kashizui committed:
Debug resolver

kashizui committed:
Remove LocalBundleClient.get_bundle_uuids

kashizui committed:
Implement wls

kashizui committed:
Fix search

kashizui committed:
Mimic checkpoint

kashizui committed:
Finish mimic

kashizui committed:
new and wrm

kashizui committed:
Remove mimic from LocalBundleClient

kashizui committed:
Remove RPC support for all commands except that used by legacy.py

kashizui committed:
Factor out XMLRPC

kashizui committed:
Copy

kashizui committed:
wadd

kashizui committed:
Fix bug

kashizui committed:
Fix bugs

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -2199,23 +2157,37 @@ def do_work_command(self, args):
             print >>self.stdout, 'Switched to worksheet %s.' % (self.worksheet_str(worksheet_info))
 
+    @staticmethod
+    def item_from_tuple(item):
```
Isn't `tuple_from_item` more accurate?

percyliang commented on codalab/lib/cli_util.py:
```diff
@@ -1,6 +1,33 @@
+from codalab.common import precondition, UsageError
+
+
+def safe_get(o, *args, **kwargs):
```
given that this lives in `cli_util.py`, the name of this function should probably be a bit more specific (contain nested_dict or something)

percyliang commented on codalab/lib/cli_util.py:
```diff
@@ -1,6 +1,33 @@
+        bundle_info.get('owner', {}).get('user_name', None)
+
+    And turns them into:
+        deep_get(bundle_info, 'owner', 'user_name')
```
what's deep_get?

percyliang commented on codalab/client/local_bundle_client.py:
```diff
@@ -12,12 +12,10 @@
 import os
```
Since we've gotten rid of the the XMLRPC server, why can't we get rid of the bundle client now?

kashizui replied:
I'm working on that :)
Currently, code in `legacy.BundleService` still uses it, but I'm looking at how much work it would be to just move all of that into `BundleService` since it's the only user.

percyliang commented on codalab/lib/worksheet_util.py:
```diff
@@ -80,6 +83,126 @@ def worksheet_line(description, uuid):
+############################################################
+
+
+class WorksheetResolver(object):
```
Hm, this makes me a bit uncomfortable.  Why do the client and server have different ways to resolve uuids?  Is the problem that in the server case, we want to make another REST API call to oneself?

Another related point is that even if the implementations were different, the functionality also seems different...it'd be nice if the functionality were the same and this was just an optimization (which we can explicitly test against).  This would make it easier to reason about.

Also, can we take this opportunity to create a new Python file that contains all the resolving logic and merge that with canonicalize.py?  The path from resolving strings to uuid starting with the CLI to the DB has been one of the most convoluted parts of the system...

kashizui replied:
@percyliang In the server case, we want to just call the model methods directly. In the client case, we have to make calls to the REST API.

I'll take a shot at cleaning this up!

kashizui replied:
@percyliang by merging with canonicalize.py, you mean we don't actually want to "create a new Python file"?

percyliang replied:
We do want to create a new file perhaps called `worksheet_resolver.py`.  We can move `canonicalize.py` into this class and delete `canonicalize.py`.

kashizui replied:
I dramatically simplified the logic. I decided to keep canonicalize, since it's not a bad name and I was able to reduce all calls on the server side to canonicalize calls.

On the client side, I just added a few additional methods in BundleCLI that make the appropriate REST API calls to resolve uuids.

percyliang reviewed (commented):
Wow, good progress!

percyliang commented:
Tried to run a job locally, and the worker seemed to have failed:

```
Traceback (most recent call last):
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 410, in _finish
    lambda: self._bundle_service.update_bundle_contents(
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 434, in _execute_bundle_service_command_with_retry
    f()
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 412, in <lambda>
    update_status))
  File "/home/pliang/research/codalab/codalab-cli/worker/bundle_service_client.py", line 21, in wrapper
    return f(*args, **kwargs)
  File "/home/pliang/research/codalab/codalab-cli/worker/bundle_service_client.py", line 150, in update_bundle_contents
    progress_callback=progress_callback)
  File "/home/pliang/research/codalab/codalab-cli/worker/rest_client.py", line 105, in _upload_with_chunked_encoding
    StringIO(response.read()))
BundleServiceException: Unable to update bundle contents in bundle service: Bad Request - remove got non-existent path:: /home/pliang/research/codalab/home/partitions/default/bundles/0xb362e13929e84069bd1786ab3d0750907r65UtZTtuC0
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 188, in _start
    self._monitor()
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 215, in _monitor
    self._worker.finish_run(self._uuid)
  File "/home/pliang/research/codalab/codalab-cli/worker/worker.py", line 228, in finish_run
    del self._runs[uuid]
KeyError: u'0xb362e13929e84069bd1786ab3d075090'
```

percyliang commented:
Never mind, this was my problem - I was running two bundle managers in
parallel by accident.  It works now.

On Sun, Oct 16, 2016 at 3:26 PM, Percy Liang percyliang@gmail.com wrote:

> Can we try it out on wstest?
> 
> On Sun, Oct 16, 2016 at 1:17 PM, Stephen Koo notifications@github.com
> wrote:
> 
> > It seems to work for me, but this is concerning.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > https://github.com/codalab/codalab-cli/pull/595#issuecomment-254071639,
> > or mute the thread
> > https://github.com/notifications/unsubscribe-auth/AAakuISM4UqTsSZvRrI3FcKQfCrfYTnbks5q0oZSgaJpZM4KXrs7
> > .

kashizui committed:
Move LocalBundleClient stuff into legacy

kashizui committed:
Remove bundlerpcserver from test-cli

kashizui committed:
Fix detach (support frozen attribute properly)

kashizui committed:
Fix mimic and copy

kashizui committed:
Update default address in circleCI

kashizui committed:
Rename safe_get -> nested_dict_get

kashizui committed:
Fix front-end bugs

kashizui committed:
Upgrade old aliases automatically

kashizui committed:
Fix more buggos

kashizui committed:
Clean up uuid resolution

kashizui committed:
Clean up local client

kashizui committed:
Fix defaults and home creation

kashizui committed:
Remove bundle client compatibility stuff

kashizui committed:
Fix missing args

kashizui committed:
Fix imports and clean up auth handlers

kashizui committed:
Clean up auth handlers

kashizui committed:
Add dummy xmlrpc endpoint

kashizui committed:
Fix canonicalize args

kashizui committed:
Fix completers

kashizui committed:
Fix last buggos??

kashizui commented:
@percyliang Should be ready for review!

All the *BundleClients, and a lot of other cruft is gone.
Given the amount of interlocking parts, I'm not sure how I could break this PR into more digestible chunks.. perhaps we could go over this some time in person.

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -38,13 +39,13 @@
     UsageError,
-    CODALAB_VERSION
+    CODALAB_VERSION,
+    NotFoundError
```
reorder

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -458,18 +458,11 @@ def simple_worksheet_str(info):
     @staticmethod
     def simple_user_str(info):
-        return '%s(%s)' % (contents_str(info.get('name')), info['id'])
+        return '%s(%s)' % (contents_str(info.get('user_name', info.get('name'))), info['id'])
```
simplify?

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -1028,81 +1049,33 @@ def copy_bundle(self, source_client, source_bundle_uuid, dest_client, dest_works
+            'detached': not add_to_worksheet,
+        })
+
+        if target_info is None:
```
Shouldn't happen?

kashizui replied:
So it turns out that this actually does happen, in the case where the bundle's contents have been deleted (using `cl rm -d`), `test-cli.py` has a check specifically for this case.

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -1028,81 +1049,33 @@ def copy_bundle(self, source_client, source_bundle_uuid, dest_client, dest_works
+
+        # Collect information about how server should unpack
+        unpack = False
+        filename = nested_dict_get(source_info, 'metadata', 'name', default='copied')
```
name always exists

kashizui commented on codalab/lib/bundle_cli.py:
```diff
@@ -1542,11 +1493,7 @@ def print_basic_info(self, client, info, raw):
-        else:
-            owner_str = '<unknown>'
-        lines.append(self.key_value_str('owner', owner_str))
+        lines.append(self.key_value_str('owner', formatting.user_str(info['owner'])))
```
Use BundleCLI.simple_user_str

kashizui committed:
Clean up

percyliang reviewed (approved):
LGTM modulo lots of further testing

kashizui commented:
Addressed comments from this morning and did a bit more cleanup, merging as soon as CircleCI approves. This will temporarily disable website upload functionality on develop/staging, until we merge codalab/codalab-worksheets#219 in too.

kashizui committed:
Fix copy empty bundle
